4.x/typescript/Recenter/README.md

```diff
@@ -0,0 +1,273 @@
+  const styles: Style = {
+    textShadow: this.state.interacting ? '-1px 0 red, 0 1px red, 1px 0 red, 0 -1px red' : ''
+  };
+  return (
+  <div
+    bind={this}
+    class={CSS.base}
+    styles={styles}
+    onclick={this._defaultCenter}>
+    <p>x: {Number(x).toFixed(3)}</p>
+    <p>y: {Number(y).toFixed(3)}</p>
+    <p>scale: {Number(scale).toFixed(5)}</p>
+  </div>
+  );
+}
+```
+
+```ts
+// Private methods
+
+private _onViewChange() {
+  let { interacting, center, scale } = this.view;
+  this.state = {
+    x: center.x,
+    y: center.y,
+    interacting,
+    scale
+  };
+}
+
+private _defaultCenter() {
+  this.view.center = this.initialCenter;
+}
+```
+
+The [render()](https://developers.arcgis.com/javascript/latest/api-reference/esri-widgets-Widget.html#render) method is the only required member of the API that must be implemented. This method must return a valid UI representation. [JSX](https://www.typescriptlang.org/docs/handbook/jsx.html) is used to define the UI. With this said, it is important to note that we are not using [React](https://reactjs.org/). The transpiled JSX is processed using a custom [JSX factory](https://developers.arcgis.com/javascript/latest/api-reference/esri-widgets-support-widget.html#tsx) therefore there is no direct equivalency between implementing a custom Widget and a React component.
+
+The snippet above creates a `style` variable of type `Style`. The `textShadow` property updates to the specified string value upon interaction. In addition, three variables: `x`, `y`, and `scale` are set to the values of `this.state`.
+
+The UI is rendered based on the specified `div` element attributes:
+
+1.  First, the `bind` attribute is set to `this`, e.g. `bind={this}`.
+2.  Next, `class` is set to the `CSS.base` value, i.e. "recenter-tool".
+3.  The `styles` will reflect the `textShadow` property set a few lines prior.
+4.  The `onclick` event (note the lower case 'c'), is set to call the private `_defaultCenter` method once the widget is clicked.
+5.  Lastly, the widget UI itself will display the `x: <value>`, `y: <value>`, and `scale: <value>` of the current view.
+
+#### Export widget
+
+At the very end of the code page, add a line to [export](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require) the object into an easily-consumable external module.
+
+```ts
+export default Recenter;
+```
+
+### 2. Compiling the TSX file
+
+Now that the widget's code is implemented, compile the _.tsx_ file to its underlying JavaScript implementation.
+
+In the command prompt, browse to the location of this sample directory and type`tsc`.
+
+This compiles any specified _.tsx_ files within the _tsconfig.json_'s files to their equivalent _.js_ files. You should now have a new _Recenter.js_ file generated in the same directory as its _.tsx_ file, in addition to a _Recenter.js.map_ [sourcemap](https://www.html5rocks.com/en/tutorials/developertools/sourcemaps/) file.
```

Comment:
I only glanced through the old html5Rocks article, it may be outdated. Consider removing the link since that article is 8+ years old.

Reply:
Updated it to something more current, https://jaxenter.com/javascript-source-maps-170126.html